extensions/amp-access/0.1/login-dialog.js

```diff
@@ -0,0 +1,185 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getMode} from '../../../src/mode';
+import {log} from '../../../src/log';
+import {parseUrl} from '../../../src/url';
+
+/** @const */
+const TAG = 'AmpAccessLogin';
+
+/** @const {!Function} */
+const assert = AMP.assert;
+
+
+/**
+ * Opens the login dialog for the specified URL. If the login dialog succeeds,
+ * the returned promised is resolved with the dialog's response. Otherwise, the
+ * returned promise is rejected.
+ * @param {!Window} win
+ * @param {string} url
+ * @return {!Promise<string>}
+ */
+export function openLoginDialog(win, url) {
+  return new LoginDialog(win, url).open();
+}
+
+
+class LoginDialog {
+  /**
+   * @param {!Window} win
+   * @param {string} url
+   */
+  constructor(win, url) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @const {string} */
+    this.url = url;
+
+    /** @private {?function(string)} */
+    this.resolve_ = null;
+
+    /** @private {?function(*)} */
+    this.reject_ = null;
+
+    /** @private {?Window} */
+    this.dialog_ = null;
+
+    /** @private {?number} */
+    this.heartbeatInterval_ = null;
+  }
+
+  /**
+   * Opens the dialog. Returns the promise that will yield with the dialog's
+   * result or will be rejected if dialog fails.
+   * @return {!Promise<string>}
+   */
+  open() {
+    assert(!this.resolve_, 'Dialog already opened');
+    return new Promise((resolve, reject) => {
+      this.resolve_ = resolve;
+      this.reject_ = reject;
+      this.openInternal_();
+    }).then(result => {
+      this.cleanup_();
+      return result;
+    }, error => {
+      this.cleanup_();
+      throw error;
+    });
+  }
+
+  /** @private */
+  cleanup_() {
+    this.resolve_ = null;
+    this.reject_ = null;
+
+    if (this.dialog_) {
+      try {
+        this.dialog_.close();
+      } catch (e) {
+        // Ignore.
+      }
+      this.dialog_ = null;
+    }
+
+    if (this.heartbeatInterval_) {
+      this.win.clearInterval(this.heartbeatInterval_);
+      this.heartbeatInterval_ = null;
+    }
+  }
+
+  /** @private */
+  openInternal_() {
+    const screen = this.win.screen;
+    const w = Math.floor(Math.min(700, screen.width * 0.9));
+    const h = Math.floor(Math.min(450, screen.height * 0.9));
+    const x = Math.floor((screen.width - w) / 2);
+    const y = Math.floor((screen.height - h) / 2);
+
+    const returnUrl = this.getReturnUrl_();
+    const returnOrigin = parseUrl(returnUrl).origin;
+    const loginUrl = this.url +
+        (this.url.indexOf('?') == -1 ? '?' : '&') +
+        'return=' + encodeURIComponent(returnUrl);
+
+    log.fine(TAG, 'Open dialog: ', loginUrl, returnOrigin, w, h, x, y);
+    this.dialog_ = this.win.open(
```

Comment:
This is definitely in a microtask. Did you try in browsers with promise polyfill whether this works successfully?

Also, could we be in a real async callback from the user's click event? I think that could happen when the CID service is not yet available?

I'd suggest to have 2 callpaths:

1. The one you have here
2. A second sync pass that only opens a window to `about:blank`. #1 sets the `location.href` to the desired URL.